jscomp/core/js_dump.ml

```diff
@@ -789,14 +800,19 @@ and expression_desc cxt ~(level : int) f x : cxt =
              [ (name_symbol, E.str p.name) ]
             else [])
         in
-        if not_is_cons = false && p.num_nonconst = 1 then tails
+        if (as_value = Some AsUnboxed || not_is_cons = false) && p.num_nonconst = 1 then tails
         else
-          ( Js_op.Lit L.tag,
-            E.str p.name
-          )
+          ( Js_op.Lit tag_name, (* TAG:xx *) 
+            match as_value with
+            | None -> E.str p.name
+            | Some as_value -> E.as_value as_value )
           :: tails
       in
-      expression_desc cxt ~level f (Object objs)
+      let exp = match objs with
+        | [(_, e)] when as_value = Some AsUnboxed -> e.expression_desc
+        | _ when as_value = Some AsUnboxed -> assert false (* should not happen *)
```

Comment:
TODO: put restriction on the variant definitions allowed, to make sure this never happens.